File: backend/src/schemas/message.py
"""
Message-related Pydantic schemas for request/response validation.

This module defines schemas for message operations including:
- Message fetch operations (start, status)
- Message retrieval (single, list with filters)
- Media download

All schemas match the OpenAPI contract in specs/001-mvp-web-api/contracts/openapi.yaml
"""

from datetime import datetime
from typing import Literal
from uuid import UUID

from fastapi import Query
from pydantic import BaseModel, Field, field_validator


# Response schemas for message fetch operations
class FetchStartedResponse(BaseModel):
    """Response when a message fetch operation is initiated."""

    fetch_id: UUID = Field(..., description="Unique fetch operation identifier")
    status: Literal["pending", "in_progress"] = Field(
        ..., description="Initial status of fetch operation"
    )
    message: str = Field(
        ...,
        example="Message fetch started. Use fetch_id to check progress.",
        description="Human-readable message",
    )


class FailedChannelInfo(BaseModel):
    """Information about a channel that failed during fetch."""

    channel_id: int = Field(..., description="Telegram channel ID that failed")
    title: str | None = Field(default=None, description="Channel title")
    error: str = Field(..., description="Error message")


class FetchStatusResponse(BaseModel):
    """Response for fetch operation status queries."""

    fetch_id: UUID = Field(..., description="Unique fetch operation identifier")
    status: Literal["pending", "in_progress", "completed", "failed", "cancelled"] = Field(
        ..., description="Current status of fetch operation"
    )
    current_phase: Literal["fetching_messages", "completed"] = Field(
        default="fetching_messages",
        description="Current phase within the fetch operation",
    )
    messages_count: int = Field(default=0, description="Number of messages fetched so far")
    started_at: datetime = Field(..., description="When the fetch started")
    completed_at: datetime | None = Field(
        default=None, description="When the fetch completed (null if not completed)"
    )
    error_message: str | None = Field(
        default=None, description="Error details (only for failed status)"
    )
    error_type: Literal["session_expired", "rate_limited", "connection_error", "unknown"] | None = (
        Field(default=None, description="Error category (only for failed status)")
    )
    # Channel-level progress fields
    total_channels: int | None = Field(
        default=None, description="Total number of channels being fetched"
    )
    completed_channels: int = Field(default=0, description="Number of channels completed so far")
    current_channel_title: str | None = Field(
        default=None, description="Title of the channel currently being fetched"
    )
    failed_channels: list[FailedChannelInfo] | None = Field(
        default=None, description="List of channels that failed during fetch"
    )
    channel_ids: list[int] | None = Field(
        default=None, description="List of channel IDs being fetched"
    )


# Response schemas for message retrieval
class MessageResponse(BaseModel):
    """Response schema for a single message."""

    id: UUID = Field(..., description="Message unique identifier")
    telegram_message_id: int = Field(..., description="Original Telegram message ID")
    chat_id: int = Field(..., description="Telegram chat/channel ID")
    chat_title: str | None = Field(default=None, description="Chat/channel name")
    chat_type: Literal["channel", "group", "supergroup", "private"] = Field(
        ..., description="Type of chat"
    )
    sender_id: int | None = Field(default=None, description="Telegram user ID of sender")
    sender_name: str | None = Field(default=None, description="Display name of sender")
    content: str | None = Field(default=None, description="Message text content")
    message_type: Literal[
        "text",
        "photo",
        "video",
        "document",
        "audio",
        "voice",
        "video_note",
        "gif",
        "sticker",
        "contact",
        "game",
        "geo",
        "poll",
        "venue",
        "other",
    ] = Field(..., description="Message type detected from Telethon properties")
    has_media: bool = Field(
        ..., description="Whether message contains media (from bool(message.media))"
    )

    # File metadata (from message.file)
    file_id: int | None = Field(default=None, description="Telegram file ID (from message.file.id)")
    file_name: str | None = Field(
        default=None, description="Original filename (from message.file.name)"
    )
    file_mime_type: str | None = Field(
        default=None, description="MIME type (from message.file.mime_type)"
    )
    file_size: int | None = Field(
        default=None, description="File size in bytes (from message.file.size)"
    )
    file_duration: int | None = Field(
        default=None,
        description="Duration in seconds for audio/video (from message.file.duration)",
    )
    file_width: int | None = Field(
        default=None,
        description="Width in pixels for images/video (from message.file.width)",
    )
    file_height: int | None = Field(
        default=None,
        description="Height in pixels for images/video (from message.file.height)",
    )

    # Message flags
    is_reply: bool = Field(..., description="Whether message is a reply (from message.is_reply)")
    is_forward: bool = Field(
        ..., description="Whether message is forwarded (from bool(message.forward))"
    )

    # Timestamps
    timestamp: datetime = Field(..., description="Original message timestamp")
    fetched_at: datetime = Field(..., description="When message was fetched")

    model_config = {"from_attributes": True}


class MessagesResponse(BaseModel):
    """Response schema for paginated message list."""

    messages: list[MessageResponse] = Field(..., description="List of messages matching filters")
    total: int = Field(..., description="Total number of messages matching filters")
    limit: int = Field(..., description="Maximum number of results per page")
    offset: int = Field(..., description="Number of results skipped")
    has_more: bool = Field(..., description="Whether there are more results available")
    search_history_clamped: dict | None = Field(
        default=None,
        description="Present when search date was clamped to plan's history limit. Contains upgrade info.",
    )


# Query parameter schema for message filtering
class MessageFilterParams(BaseModel):
    """Query parameters for filtering and searching messages.

    Supports full-text search, chat filtering, date range, message type, and media filtering.
    All parameters are optional and can be combined.
    """

    search: str | None = Field(
        default=None,
        description="Full-text search query for message content",
        min_length=1,
        max_length=500,
    )
    chat_ids: list[int] | None = Query(
        default=None, description="Filter by specific chat IDs (can specify multiple)"
    )
    start_date: datetime | None = Field(
        default=None,
        description="Filter messages after this date (inclusive)",
    )
    end_date: datetime | None = Field(
        default=None,
        description="Filter messages before this date (inclusive)",
    )
    message_type: (
        Literal[
            "text",
            "photo",
            "video",
            "document",
            "audio",
            "voice",
            "video_note",
            "gif",
            "sticker",
            "contact",
            "game",
            "geo",
            "poll",
            "venue",
            "other",
        ]
        | None
    ) = Field(
        default=None,
        description="Filter by message type",
    )
    has_media: bool | None = Field(
        default=None,
        description="Filter by presence of media (true=media only, false=no media)",
    )
    limit: int = Field(
        default=50,
        ge=1,
        le=200,
        description="Maximum number of results (1-200)",
    )
    offset: int = Field(
        default=0,
        ge=0,
        description="Number of results to skip for pagination",
    )

    @field_validator("end_date")
    @classmethod
    def validate_date_range(cls, end_date: datetime | None, info) -> datetime | None:
        """Validate that end_date is after start_date if both are provided."""
        start_date = info.data.get("start_date")
        if start_date is not None and end_date is not None:
            if end_date < start_date:
                raise ValueError("end_date must be after or equal to start_date")
        return end_date


class ChatFilterParams(BaseModel):
    search: str | None = Field(default=None, max_length=100)
    limit: int = Field(default=200, ge=1, le=500)
    offset: int = Field(default=0, ge=0)


class ChatResponse(BaseModel):
    chat_id: str
    title: str
    type: Literal["channel", "group", "supergroup", "private"]
    message_count: int


class ChatsResponse(BaseModel):
    chats: list[ChatResponse]
    total: int
    limit: int
    offset: int
    has_more: bool


class MessageIdsResponse(BaseModel):
    """Response schema for message IDs only (lightweight endpoint)."""

    ids: list[UUID] = Field(..., description="List of message UUIDs matching filters")
    total: int = Field(..., description="Total number of messages matching filters")
    search_history_clamped: dict | None = Field(
        default=None,
        description="Present when search date was clamped to plan's history limit. Contains upgrade info.",
    )

File: frontend/src/types/message.ts
/**
 * Message types based on backend API contracts
 * Source: specs/003-minimal-frontend-webapp/contracts/messages-contract.md
 */

/**
 * DEPRECATED: Backend returns flat file_* fields, not nested media object
 * Keeping for backwards compatibility with tests
 */
export interface MediaInfo {
  /** Media type: photo, video, document, audio, voice, etc. */
  type: 'photo' | 'video' | 'document' | 'audio' | 'voice' | 'video_note' | 'sticker'

  /** File name (for documents) */
  fileName?: string

  /** File size in bytes */
  fileSize?: number

  /** MIME type */
  mimeType?: string

  /** Image/video dimensions */
  dimensions?: {
    width: number
    height: number
  }

  /** Video/audio duration in seconds */
  duration?: number
}

export interface Message {
  /** Unique message ID (UUID) */
  id: string

  /** User ID who owns this message */
  userId: string

  /** Telegram message ID (for deduplication) */
  telegramMessageId: number

  /** Message text content (may be null for media-only messages) */
  content: string | null

  /** Sender information */
  sender: {
    /** Sender's Telegram user ID */
    telegramUserId: string

    /** Sender's display name */
    name: string
  }

  /** Chat/channel information */
  chat: {
    /** Chat ID */
    telegramChatId: string

    /** Chat title */
    title: string

    /** Chat type: private, group, supergroup, channel */
    type: 'private' | 'group' | 'supergroup' | 'channel'
  }

  /** Message timestamp (ISO 8601 format) */
  timestamp: string

  /** Message type (text, photo, video, document, etc.) */
  type: string

  /** Media metadata (if message contains media) */
  media: MediaInfo | null

  /** Is this message a reply to another message? */
  isReply: boolean

  /** Is this message forwarded from another chat? */
  isForwarded: boolean

  /** ISO timestamp when this message was fetched from Telegram */
  fetchedAt: string
}

export interface FetchOperation {
  /** Unique fetch operation ID (UUID) */
  id: string

  /** User ID who initiated the fetch */
  userId: string

  /** Fetch status */
  status: 'pending' | 'in_progress' | 'completed' | 'failed' | 'cancelled'

  /** Current phase within the fetch operation */
  currentPhase?: 'fetching_messages' | 'processing_embeddings' | 'completed'

  /** Number of messages fetched so far */
  messageCount: number

  /** Error message (if status is 'failed') */
  error: string | null

  /** ISO timestamp when fetch was started */
  startedAt: string

  /** ISO timestamp when fetch completed/failed (null if still running) */
  completedAt: string | null

  /** Total number of channels to fetch from */
  totalChannels?: number

  /** Number of channels completed */
  completedChannels?: number

  /** Current channel being processed */
  currentChannelTitle?: string

  /** Failed channels with error details */
  failedChannels?: Array<{
    channel_id: number
    title: string | null
    error: string
  }>

  /** Channel IDs being fetched (if specified) */
  channelIds?: number[]
}

export interface MessageFilters {
  /** Text search query (searches message content) */
  search?: string

  /** Filter by specific chat ID */
  chatId?: string

  /** Filter by start date (ISO 8601 date string, e.g., "2025-01-01") */
  startDate?: string

  /** Filter by end date (ISO 8601 date string) */
  endDate?: string

  /** Filter by message type (e.g., "photo", "video") */
  type?: string

  /** Show only messages with media (true/false as string) */
  hasMedia?: string

  /** Current page number (for pagination) */
  page?: string

  /** Number of results per page (default: 50) */
  limit?: string
}

export interface PaginationMeta {
  /** Current page number (zero-indexed) */
  currentPage: number

  /** Number of items per page */
  pageSize: number

  /** Total number of items across all pages */
  totalItems: number

  /** Total number of pages */
  totalPages: number

  /** Is there a next page available? */
  hasNextPage: boolean

  /** Is there a previous page available? */
  hasPrevPage: boolean
}

// API response types (these are what the FRONTEND receives, converted from backend snake_case)
export interface TriggerFetchResponse {
  fetchId: string
  status: string
  message?: string
}

export interface FetchStatusResponse {
  fetch_id: string
  status: 'pending' | 'in_progress' | 'completed' | 'failed' | 'cancelled'
  current_phase?: 'fetching_messages' | 'processing_embeddings' | 'completed'
  messages_count: number
  error_message: string | null
  error_type?: 'session_expired' | 'rate_limited' | 'connection_error' | 'unknown' | null
  started_at: string
  completed_at: string | null
  total_channels?: number
  completed_channels?: number
  current_channel_title?: string
  failed_channels?: Array<{
    channel_id: number
    title: string | null
    error: string
  }>
  channel_ids?: number[]
}

export interface GetMessagesResponse {
  messages: Array<{
    id: string
    user_id: string
    telegram_message_id: number
    content: string | null
    sender_id: number | null
    sender_name: string | null
    chat_id: number
    chat_title: string
    chat_type: string
    timestamp: string
    message_type: string
    has_media: boolean
    // Flat file metadata fields (from backend MessageResponse schema)
    file_id: number | null
    file_name: string | null
    file_mime_type: string | null
    file_size: number | null
    file_duration: number | null
    file_width: number | null
    file_height: number | null
    is_reply: boolean
    is_forward: boolean
    fetched_at: string
  }>
  total: number
  offset: number
  limit: number
  has_more: boolean
}

export interface GetMessageDetailResponse {
  id: string
  user_id: string
  telegram_message_id: number
  content: string | null
  sender_id: number | null
  sender_name: string | null
  chat_id: number
  chat_title: string
  chat_type: string
  timestamp: string
  message_type: string
  has_media: boolean
  // Flat file metadata fields (from backend MessageResponse schema)
  file_id: number | null
  file_name: string | null
  file_mime_type: string | null
  file_size: number | null
  file_duration: number | null
  file_width: number | null
  file_height: number | null
  is_reply: boolean
  is_forward: boolean
  fetched_at: string
}

export interface MediaDownloadResponse {
  // Binary data - handled by browser download
  data: Blob
  filename: string
  mimeType: string
}

/**
 * Message loading state for UI indicators
 * Used by useMessages hook to determine which loading indicator to show
 */
export type MessageLoadingState =
  | 'idle' // No operation in progress
  | 'loading' // Initial load (show skeleton cards)
  | 'refreshing' // Manual refresh (show refresh spinner)
  | 'loadingMore' // Pagination (show button spinner)
  | 'error' // Error state (show error message)
